File: mobile/lib/models/progress.dart
class Progress {
  final String id;
  final String userId;
  final double? bmi;
  final double? caloriesIntake;
  final double? calorieDeficit;
  final DateTime date;
  final DateTime? lastBmiUpdate;
  final DateTime? lastCaloriesUpdate;

  Progress({
    required this.id,
    required this.userId,
    this.bmi,
    this.caloriesIntake,
    this.calorieDeficit,
    DateTime? date,
    this.lastBmiUpdate,
    this.lastCaloriesUpdate,
  }) : date = date ?? DateTime.now();

  factory Progress.fromJson(Map<String, dynamic> json) {
    return Progress(
      id: json['_id'] ?? '',
      userId: json['userId'] ?? json['user'] ?? '',
      bmi: (json['bmi'] as num?)?.toDouble(),
      caloriesIntake: (json['caloriesIntake'] as num?)?.toDouble(),
      calorieDeficit: (json['calorieDeficit'] as num?)?.toDouble(),
      date: json['date'] != null ? DateTime.parse(json['date']) : null,
      lastBmiUpdate: json['lastBmiUpdate'] != null ? DateTime.parse(json['lastBmiUpdate']) : null,
      lastCaloriesUpdate: json['lastCaloriesUpdate'] != null ? DateTime.parse(json['lastCaloriesUpdate']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'userId': userId,
      if (bmi != null) 'bmi': bmi,
      if (caloriesIntake != null) 'caloriesIntake': caloriesIntake,
      if (calorieDeficit != null) 'calorieDeficit': calorieDeficit,
      'date': date.toIso8601String(),
      if (lastBmiUpdate != null) 'lastBmiUpdate': lastBmiUpdate!.toIso8601String(),
      if (lastCaloriesUpdate != null) 'lastCaloriesUpdate': lastCaloriesUpdate!.toIso8601String(),
    };
  }

  // Helper method to check if BMI can be updated (7 days)
  bool canUpdateBmi() {
    if (lastBmiUpdate == null) return true;
    final daysSince = DateTime.now().difference(lastBmiUpdate!).inDays;
    return daysSince >= 7;
  }

  // Helper method to check if Calories can be updated (24 hours)
  bool canUpdateCalories() {
    if (lastCaloriesUpdate == null) return true;
    final hoursSince = DateTime.now().difference(lastCaloriesUpdate!).inHours;
    return hoursSince >= 24;
  }

  // Get days until next BMI update
  int daysUntilNextBmiUpdate() {
    if (lastBmiUpdate == null) return 0;
    final daysSince = DateTime.now().difference(lastBmiUpdate!).inDays;
    return (7 - daysSince).clamp(0, 7);
  }

  // Get hours until next Calories update
  int hoursUntilNextCaloriesUpdate() {
    if (lastCaloriesUpdate == null) return 0;
    final hoursSince = DateTime.now().difference(lastCaloriesUpdate!).inHours;
    return (24 - hoursSince).clamp(0, 24);
  }
}

File: backend/models/Progress.js
const mongoose = require('mongoose');

const ProgressSchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  bmi: {
    type: Number,
    max: 300
  },
  caloriesIntake: {
    type: Number
  },
  calorieDeficit: {
    type: Number
  },
  weight: {
    type: Number,
    min: 20,
    max: 300
  },
  bodyFatPercentage: {
    type: Number,
    min: 0,
    max: 100
  },
  muscleMass: {
    type: Number,
    min: 0
  },
  // Separate tracking for last update times
  lastBmiUpdate: {
    type: Date
  },
  lastCaloriesUpdate: {
    type: Date
  },
  date: {
    type: Date,
    default: Date.now
  }
}, {
  timestamps: true
});

// Index for efficient queries
ProgressSchema.index({ user: 1, createdAt: -1 });

// Method to check if BMI can be updated (7 days restriction)
ProgressSchema.methods.canUpdateBmi = function() {
  if (!this.lastBmiUpdate) return true;
  
  const now = new Date();
  const daysSinceLastUpdate = (now - this.lastBmiUpdate) / (1000 * 60 * 60 * 24);
  return daysSinceLastUpdate >= 7;
};

// Method to check if Calories can be updated (24 hours restriction)
ProgressSchema.methods.canUpdateCalories = function() {
  if (!this.lastCaloriesUpdate) return true;
  
  const now = new Date();
  const hoursSinceLastUpdate = (now - this.lastCaloriesUpdate) / (1000 * 60 * 60);
  return hoursSinceLastUpdate >= 24;
};

// Method to get days until next BMI update
ProgressSchema.methods.daysUntilNextBmiUpdate = function() {
  if (!this.lastBmiUpdate) return 0;
  
  const now = new Date();
  const daysSinceLastUpdate = (now - this.lastBmiUpdate) / (1000 * 60 * 60 * 24);
  return Math.max(0, Math.ceil(7 - daysSinceLastUpdate));
};

// Method to get hours until next Calories update
ProgressSchema.methods.hoursUntilNextCaloriesUpdate = function() {
  if (!this.lastCaloriesUpdate) return 0;
  
  const now = new Date();
  const hoursSinceLastUpdate = (now - this.lastCaloriesUpdate) / (1000 * 60 * 60);
  return Math.max(0, Math.ceil(24 - hoursSinceLastUpdate));
};

module.exports = mongoose.model('Progress', ProgressSchema);
